avocado_vt/plugins/vt_runner.py

```diff
@@ -0,0 +1,279 @@
+            env_filename = os.path.join(data_dir.get_tmp_dir(),
+                                        params.get("env", "env"))
+            env = utils_env.Env(env_filename, self.env_version)
+
+            test_passed = False
+            t_type = None
+
+            try:
+                try:
+                    # Pre-process
+                    try:
+                        params = env_process.preprocess(self, params, env)
+                    finally:
+                        self._safe_env_save(env)
+
+                    # Run the test function
+                    for t_type in t_types:
+                        test_module = test_modules[t_type]
+                        run_func = utils_misc.get_test_entrypoint_func(
+                            t_type, test_module)
+                        try:
+                            run_func(self, params, env)
+                            self.verify_background_errors()
+                        finally:
+                            self._safe_env_save(env)
+                    test_passed = True
+                    error_message = funcatexit.run_exitfuncs(env, t_type)
+                    if error_message:
+                        raise exceptions.TestWarn("funcatexit failed with: %s" %
+                                                  error_message)
+
+                except:  # nopep8 Old-style exceptions are not inherited from Exception()
+                    stacktrace.log_exc_info(sys.exc_info(), 'avocado.test')
+                    if t_type is not None:
+                        error_message = funcatexit.run_exitfuncs(env, t_type)
+                        if error_message:
+                            self.log.error(error_message)
+                    try:
+                        env_process.postprocess_on_error(self, params, env)
+                    finally:
+                        self._safe_env_save(env)
+                    raise
+
+            finally:
+                # Post-process
+                try:
+                    try:
+                        params['test_passed'] = str(test_passed)
+                        env_process.postprocess(self, params, env)
+                    except:  # nopep8 Old-style exceptions are not inherited from Exception()
+
+                        stacktrace.log_exc_info(sys.exc_info(),
+                                                'avocado.test')
+                        if test_passed:
+                            raise
+                        self.log.error("Exception raised during postprocessing:"
+                                       " %s", sys.exc_info()[1])
+                finally:
+                    if self._safe_env_save(env) or params.get("env_cleanup",
+                                                              "no") == "yes":
+                        env.destroy()  # Force-clean as it can't be stored
+
+        except Exception as e:
+            try:
+                if params.get("abort_on_error") != "yes":
```

Comment:
In the current refactoring `params` might exist in case of early exception here, we have to provide it before the try clause.

Reply:
True, I will fix it, thanks

Reply:
There is a similar problem with `env` here if an exception is hit. I would recommend going through the entire block and making sure there are no uninitialized variables since this refactoring moved the block to a location where a lot of them might no longer exist.